servers/rendering/shader_language.cpp

```diff
@@ -4991,55 +4952,104 @@ ShaderLanguage::Node *ShaderLanguage::_parse_expression(BlockNode *p_block, cons
+											}
+
+											if (error) {
+												_set_error(vformat("Varying '%s' cannot be passed for the '%s' parameter in that context!", varname, _get_qualifier_str(arg_qual)));
+												return nullptr;
+											}
+										}
+									}
+
+									bool is_const_arg = call_function->arguments[i].is_const;
+
+									if (is_const_arg || is_out_arg) {
+										StringName varname;
+
+										if (n->type == Node::TYPE_CONSTANT || n->type == Node::TYPE_OPERATOR || n->type == Node::TYPE_ARRAY_CONSTRUCT) {
+											if (!is_const_arg) {
 												error = true;
 											}
 										} else if (n->type == Node::TYPE_ARRAY) {
 											ArrayNode *an = static_cast<ArrayNode *>(n);
-											if (an->call_expression != nullptr || an->is_const) {
+											if (!is_const_arg && (an->call_expression != nullptr || an->is_const)) {
 												error = true;
 											}
+											varname = an->name;
 										} else if (n->type == Node::TYPE_VARIABLE) {
 											VariableNode *vn = static_cast<VariableNode *>(n);
-											if (vn->is_const) {
+											if (vn->is_const && !is_const_arg) {
 												error = true;
-											} else {
-												StringName varname = vn->name;
-												if (shader->constants.has(varname)) {
-													error = true;
-												} else if (shader->uniforms.has(varname)) {
-													error = true;
-												} else {
-													if (shader->varyings.has(varname)) {
-														_set_error(vformat("Varyings cannot be passed for '%s' parameter!", _get_qualifier_str(call_function->arguments[i].qualifier)));
-														return nullptr;
-													}
-													if (p_function_info.built_ins.has(varname)) {
-														BuiltInInfo info = p_function_info.built_ins[varname];
-														if (info.constant) {
-															error = true;
-														}
-													}
-												}
 											}
+											varname = vn->name;
 										} else if (n->type == Node::TYPE_MEMBER) {
 											MemberNode *mn = static_cast<MemberNode *>(n);
-											if (mn->basetype_const) {
+											if (mn->basetype_const && is_out_arg) {
 												error = true;
 											}
 										}
+										if (!error && varname != StringName()) {
+											if (shader->constants.has(varname)) {
+												error = true;
+											} else if (shader->uniforms.has(varname)) {
+												error = true;
+											} else if (p_function_info.built_ins.has(varname)) {
+												BuiltInInfo info = p_function_info.built_ins[varname];
+												if (info.constant) {
+													error = true;
+												}
+											}
+										}
+
 										if (error) {
-											_set_error(vformat("Constant value cannot be passed for '%s' parameter!", _get_qualifier_str(call_function->arguments[i].qualifier)));
+											_set_error(vformat("Constant value cannot be passed for '%s' parameter!", _get_qualifier_str(arg_qual)));
 											return nullptr;
 										}
```

Comment:
Also here, I've detected and fixed several bugs when passing `const` parameters to the functions. For example:
```
void foo(const int t) {
}

void fragment() {
const int t = 0;
foo(t); // cannot be passed before fix
}
```